src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java

```diff
@@ -1,88 +1,322 @@
- * <br>
- * This event is fired on the {@link MinecraftForge#EVENT_BUS}.
+ * All subclasses are fired on {@link MinecraftForge#EVENT_BUS}.
+ * See the individual documentation on each subevent for more details.
  **/
-@Cancelable
 public class PlayerInteractEvent extends PlayerEvent
 {
-    public static enum Action
+    private final EnumHand hand;
+    private final ItemStack stack;
+    private final BlockPos pos;
+    private final EnumFacing face;
+
+    private PlayerInteractEvent(EntityPlayer player, EnumHand hand, ItemStack stack, BlockPos pos, EnumFacing face)
     {
-        RIGHT_CLICK_AIR,
-        RIGHT_CLICK_BLOCK,
-        LEFT_CLICK_BLOCK
+        super(Preconditions.checkNotNull(player, "Null player in PlayerInteractEvent!"));
+        this.hand = Preconditions.checkNotNull(hand, "Null hand in PlayerInteractEvent!");
+        this.stack = stack;
+        this.pos = Preconditions.checkNotNull(pos, "Null position in PlayerInteractEvent!");
+        this.face = face;
     }
 
-    private final Action action;
-    private final World world;
-    private final BlockPos pos;
-    private final EnumFacing face; // Can be null if unknown
-    private final Vec3d localPos; // Can be null if unknown
+    /**
+     * This event is fired on both sides whenever a player interacts with an entity.
+     * It has the local position (where on the entity) the player interacted at.
+     * Canceling the event clientside will cause processing to continue to EntityInteract (below),
+     * while canceling serverside will simply do no further processing.
+     */
+    @Cancelable
+    public static class EntityInteractSpecific extends PlayerInteractEvent
+    {
+        private final Vec3d localPos;
+        private final Entity target;
+
+        public EntityInteractSpecific(EntityPlayer player, EnumHand hand, ItemStack stack, Entity target, Vec3d localPos)
+        {
+            super(player, hand, stack, new BlockPos(target), null);
+            this.localPos = localPos;
+            this.target = target;
+        }
 
-    private Result useBlock = DEFAULT;
-    private Result useItem = DEFAULT;
+        /**
+         * Returns the local interaction position. This is a 3D vector, where (0, 0, 0) is centered exactly at the
+         * center of the entity's bounding box at their feet. This means the X and Z values will be in the range
+         * [-width / 2, width / 2] while Y values will be in the range [0, height]
+         * @return The local position
+         */
+        public Vec3d getLocalPos()
+        {
+            return localPos;
+        }
 
-    @Deprecated
-    public PlayerInteractEvent(EntityPlayer player, Action action, BlockPos pos, EnumFacing face, World world)
+        public Entity getTarget()
+        {
+            return target;
+        }
+    }
+
+    /**
+     * This event is fired on both sides after EntityInteractSpecific is fired above.
+     * If the above event is canceled, this event will still fire.
```

Comment:
Why would this event fire? What is the case where it fires, and what is it actually signifying that the previous event was cancelled? Is this more of a notification event instead of "oh Item.onItemUse is going to be called".